src/pyclass.rs

```diff
@@ -17,7 +18,14 @@ use std::{
 ///
 /// The `#[pyclass]` attribute automatically implements this trait for your Rust struct,
 /// so you normally don't have to use this trait directly.
-pub trait PyClass:
+///
+/// # Safety
+///
+/// If `T` implements [`MutablePyClass`], then implementations must override the default methods.
+///
+/// If `T`  does not implement [`MutablePyClass`], then implementations should not  override the
+/// default methods.
+pub unsafe trait PyClass:
```

Comment:
>  The contract between the relationship of MutablePyClass & PyClass is a bit scary.
>
>  I'd rather keep these details out of the public API, because I expect we'll need to iterate on PyClass a few times yet and it would be nice if fewer users broke / we didn't have to restrict ourselves to iterating in semver-breaking releases.

I mean, how public is `PyClass` actually? Are people really out there implementing `PyClass` for their structs? They also need to implement "hidden" traits to do so. I'd like to think `PyClass` is read only public, so adding things shouldn't be an issue.

That said, I agree this coupling between the two traits is not ideal. In particular `PyClass`' trait methods don't feel like a natural fit.

But I like the PyCell api that we end up with, in particular that it doesn't add a bunch of trait bounds to it. It could also be implemented cleanly with specialization or some sort of negative trait bounds, but we don't have those.

>I wonder if there's a possible design where we introduce multiple types of BorrowFlag, similar to how we have the ThreadChecker / ThreadCheckerStub?

This would involve a runtime check, right?

> I'm thinking it might be possible to write an implementation where an "immutable" borrowflag always fails to borrow mutably and is a no-op when borrowing immutably.

Would that be be a runtime error or a compilation error? I really prefer the latter, if possible.

Reply:
> I mean, how public is `PyClass` actually? Are people really out there implementing `PyClass` for their structs? They also need to implement "hidden" traits to do so. I'd like to think `PyClass` is read only public, so adding things shouldn't be an issue.

That's true, however everything the macros access is "public" in some form. It's not really about users implementing `PyClass`, but more the existence of these methods - I would rather they were in a place where its obvious that users are not expected to call the methods. I've been making `PyClassImpl` the bucket of "stuff which we can freely change so don't blame us if you use it and it breaks in a patch release".

> This would involve a runtime check, right?

Yes, but if the immutable borrowflag is trivial no-op the call should be optimized away to nothing in release builds.

> Would that be be a runtime error or a compilation error? I really prefer the latter, if possible.

Agreed compilation error would be best if we can make it so. I do like the design you've built for `Py` and `PyCell` here where it's impossible to even try to borrow mutably if the class is not mutable.

I'll try and prepare a PR for you to add to this branch the idea I'm thinking a bit better...

Reply:
>  I would rather they were in a place where its obvious that users are not expected to call the methods. I've been making PyClassImpl the bucket of "stuff which we can freely change so don't blame us if you use it and it breaks in a patch release".

Yes, that looks like a better place for them.

> I'm thinking it might be possible to write an implementation where an "immutable" borrowflag always fails to borrow mutably and is a no-op when borrowing immutably.

Now that I've given it a second readover - do you mean that rather than the actual borrowing itself, we proxy the borrowflag checking through a trait? That is probably more appropriate I think.

Reply:
Yes exactly - like we have the `PyClassThreadChecker` trait for `#[unsendable]`, we could have a `PyClassBorrowChecker` trait for this use-case.

Reply:
(I think it's going to be a couple of days yet before I'll have a chance to write the PR, if you think you've got enough to go on feel free to go with it?)

Reply:
I think I can work with that.